grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java

```diff
@@ -67,285 +63,321 @@
+        return map1.entries()
+                .stream()
+                .collect(flatteningToMultimap(Map.Entry::getKey, e -> map2.get(e.getValue()).stream()));
+    }
 
-        visitedVars.push(name);
-        Concept concept = concepts.computeIfAbsent(name, n -> addConcept(var));
-        assert concept != null : var ;
-        if (var.getVarName().isUserDefinedName()) namedConcepts.put(name, concept);
-        visitedVars.pop();
-        return concept;
+    private Answer insertAll(Answer results) {
+        concepts.putAll(results.map());
+
+        sortProperties().forEach(property -> property.insert(this));
+
+        conceptBuilders.forEach((var, builder) -> concepts.put(var, builder.build()));
+
+        Map<Var, Concept> namedConcepts = Maps.filterKeys(concepts, Var::isUserDefinedName);
+        return new QueryAnswer(namedConcepts);
     }
 
     /**
-     * @param varToAdd the {@link VarPatternAdmin} that is to be added into the graph
-     * @return the concept representing the given {@link VarPatternAdmin}, creating it if it doesn't exist
+     * Produce a valid ordering of the properties by using the given dependency information.
+     *
+     * <p>
+     *     This method uses a topological sort (Kahn's algorithm) in order to find a valid ordering.
+     * </p>
      */
-    private Concept addConcept(VarPatternAdmin varToAdd) {
-        VarPatternAdmin var = mergeVar(varToAdd);
+    private ImmutableList<VarAndProperty> sortProperties() {
+        ImmutableList.Builder<VarAndProperty> sorted = ImmutableList.builder();
 
-        Optional<IsaProperty> type = var.getProperty(IsaProperty.class);
-        Optional<SubProperty> sub = var.getProperty(SubProperty.class);
+        // invertedDependencies is intended to just be a 'view' on dependencies, so when dependencies is modified
+        // we should always also modify invertedDependencies (and vice-versa).
+        Multimap<VarAndProperty, VarAndProperty> dependencies = HashMultimap.create(this.dependencies);
+        Multimap<VarAndProperty, VarAndProperty> invertedDependencies = HashMultimap.create();
+        Multimaps.invertFrom(dependencies, invertedDependencies);
 
-        if (type.isPresent() && sub.isPresent()) {
-            String printableName = var.getPrintableName();
-            throw GraqlQueryException.insertIsaAndSub(printableName);
-        }
+        Queue<VarAndProperty> propertiesWithoutDependencies =
+                new ArrayDeque<>(Sets.filter(properties, property -> dependencies.get(property).isEmpty()));
 
-        Optional<Label> typeLabel = var.getTypeLabel();
-        Optional<ConceptId> id = var.getId();
+        VarAndProperty property;
 
-        typeLabel.ifPresent(label -> {
-            if (type.isPresent()) {
-                throw GraqlQueryException.insertInstanceWithLabel(label);
-            }
-        });
+        while ((property = propertiesWithoutDependencies.poll()) != null) {
```

Comment:
Just checking my understanding here for this specific loop:

You basically iteratively filtering out dependencies by starting with with things which have no dependencies. I.e. things which can be inserted. When you know what these things are you then eliminate them as dependencies. This in turn lets you find new things with no dependencies and then the process repeats itself?

If my above description is correct than stating this in a comment may have value.